internal/controllers/objectsetphases/phase_reconciler.go

```diff
@@ -0,0 +1,506 @@
+	}
+
+	// Ensure to watch this type of object, also during teardown!
+	// If the controller was restarted or crashed during deletion, we might not have a cache in memory anymore.
+	if err := r.dynamicCache.Watch(
+		ctx, owner, desiredObj); err != nil {
+		return false, fmt.Errorf("watching new resource: %w", err)
+	}
+
+	currentObj := desiredObj.DeepCopy()
+	err = r.dynamicCache.Get(
+		ctx, client.ObjectKeyFromObject(desiredObj), currentObj)
+	if err != nil && errors.IsNotFound(err) {
+		// No matter who the owner of this object is,
+		// it's already gone.
+		return true, nil
+	}
+	if err != nil {
+		return false, fmt.Errorf("getting object for teardown: %w", err)
+	}
+
+	if !r.ownerStrategy.IsController(owner, currentObj) {
+		// this object is owned by someone else
+		// so we don't have to delete it for cleanup,
+		// but we still want to remove ourself as owner.
+		r.ownerStrategy.RemoveOwner(owner, currentObj)
+		if err := r.writer.Update(ctx, currentObj); err != nil {
+			return false, fmt.Errorf("removing owner reference: %w", err)
+		}
+		return true, nil
+	}
+
+	err = r.writer.Delete(ctx, currentObj)
+	if err != nil && errors.IsNotFound(err) {
+		return true, nil
+	}
+	if err != nil {
+		return false, fmt.Errorf("deleting object for teardown: %w", err)
+	}
+
+	return false, nil
+}
+
+func (r *PhaseReconciler) reconcilePhaseObject(
+	ctx context.Context, owner PhaseObjectOwner,
+	phaseObject corev1alpha1.ObjectSetObject,
+	previous []client.Object,
+) (actualObj *unstructured.Unstructured, err error) {
+	desiredObj, err := r.desiredObject(
+		ctx, owner.ClientObject(), phaseObject)
+	if err != nil {
+		return nil, fmt.Errorf("building desired object: %w", err)
+	}
+
+	// Ensure to watch this type of object.
+	if err := r.dynamicCache.Watch(
+		ctx, owner.ClientObject(), desiredObj); err != nil {
+		return nil, fmt.Errorf("watching new resource: %w", err)
+	}
+
+	if owner.IsPaused() {
+		actualObj = desiredObj.DeepCopy()
+		if err := r.dynamicCache.Get(ctx, client.ObjectKeyFromObject(desiredObj), actualObj); err != nil {
+			return nil, fmt.Errorf("looking up object while paused: %w", err)
+		}
+		return actualObj, nil
+	}
+
+	return r.reconcileObject(ctx, owner, desiredObj, previous)
+}
+
+// Builds an object as specified in a phase.
+// Includes system labels, namespace and owner reference.
+func (r *PhaseReconciler) desiredObject(
+	ctx context.Context, owner client.Object,
+	phaseObject corev1alpha1.ObjectSetObject,
+) (desiredObj *unstructured.Unstructured, err error) {
+	desiredObj, err = unstructuredFromObjectSetObject(&phaseObject)
+	if err != nil {
+		return nil, err
+	}
+
+	// Default namespace to the owners namespace
+	if len(desiredObj.GetNamespace()) == 0 {
+		desiredObj.SetNamespace(
+			owner.GetNamespace())
+	}
+
+	// Set cache label
+	labels := desiredObj.GetLabels()
+	if labels == nil {
+		labels = map[string]string{}
+	}
+	labels[controllers.DynamicCacheLabel] = "True"
+	desiredObj.SetLabels(labels)
+
```

Comment:
Don't we also want to set the owner's revision in the desired object ?

Reply:
:100: You are right, adding that now :+1:

Reply:
fixed!